src/addonloader/luaaddonstate.cpp

```diff
@@ -128,46 +128,107 @@ std::tuple<> LuaAddonState::logImpl(const char *msg) {
     return {};
 }
 
-std::tuple<int> LuaAddonState::watchEventImpl(const char *event,
-                                              const char *function) {
-    std::string_view eventView(event);
-    bool validEvent = true;
-    EventType type;
-    if (eventView == "KeyEvent") {
-        type = EventType::InputContextKeyEvent;
-    } else {
-        validEvent = false;
-    }
-    if (!validEvent) {
-        throw std::runtime_error("Invalid eventype");
-    }
-    int newId = ++currentId_;
-    std::unique_ptr<HandlerTableEntry<EventHandler>> handler;
-    handler = instance_->watchEvent(
-        type, EventWatcherPhase::PreInputMethod, [this, newId](Event &event) {
-            auto iter = eventHandler_.find(newId);
+template <typename T>
+std::unique_ptr<HandlerTableEntry<EventHandler>> LuaAddonState::watchEvent(
+    EventType type, int id,
+    std::function<int(std::unique_ptr<LuaState> &, T &)> pushArguments,
+    std::function<void(std::unique_ptr<LuaState> &, T &)> handleReturnValue) {
+    return instance_->watchEvent(
+        type, EventWatcherPhase::PreInputMethod,
+        [this, id, pushArguments, handleReturnValue](Event &event_) {
+            auto iter = eventHandler_.find(id);
+            int argc = 0;
             if (iter == eventHandler_.end()) {
                 return;
             }
-            auto &keyEvent = static_cast<KeyEvent &>(event);
-            ScopedICSetter setter(inputContext_,
-                                  keyEvent.inputContext()->watch());
+            auto &event = static_cast<T &>(event_);
+            ScopedICSetter setter(inputContext_, event.inputContext()->watch());
             state_->lua_getglobal(iter->second.function().data());
-            state_->lua_pushinteger(keyEvent.key().sym());
-            state_->lua_pushinteger(keyEvent.key().states());
-            state_->lua_pushboolean(keyEvent.isRelease());
-            int rv = state_->lua_pcallk(3, 1, 0, 0, nullptr);
+            if (pushArguments) {
+                argc = pushArguments(state_, event);
+            }
+            int rv = state_->lua_pcallk(argc, 1, 0, 0, nullptr);
             if (rv != 0) {
                 LuaPError(rv, "lua_pcall() failed");
                 LuaPrintError(*this);
             } else if (state_->lua_gettop() >= 1) {
-                auto b = state_->lua_toboolean(-1);
-                if (b) {
-                    keyEvent.filterAndAccept();
+                if (handleReturnValue) {
+                    handleReturnValue(state_, event);
                 }
             }
             state_->pop(state_->lua_gettop());
         });
+}
+
+std::tuple<int> LuaAddonState::watchEventImpl(int eventType,
+                                              const char *function) {
+    int newId = currentId_ + 1;
+    std::unique_ptr<HandlerTableEntry<EventHandler>> handler = nullptr;
+
+    switch (static_cast<EventType>(eventType)) {
+    case EventType::InputContextCreated:
```

Comment:
These case can be merged into a same branch InputContextEvent

Reply:
Updated